kcmdline.c

```diff
@@ -96,13 +97,202 @@ EFI_STATUS check_cmdline(CONST CHAR8 *cmdline, UINTN cmdline_len) {
+			}
+		} else {
+			// builtin has a marker, check that only one.
+			if (strstra(p + marker_len, marker) != NULL) {
+				status = EFI_INVALID_PARAMETER;
+				UnicodeSPrint(errbuf, errbuf_buflen,
+						L"%a XXXX appears more than once in builtin cmdline\n", marker);
+				goto out;
+			}
+
+			part1_len = (p - builtin);
+			part2_len = builtin_len - marker_len - part1_len;
+			if (!((p == builtin || *(p-1) == ' ') &&
+					(part2_len == 0 || *(p + marker_len) == ' '))) {
+				status = EFI_INVALID_PARAMETER;
+				UnicodeSPrint(errbuf, errbuf_buflen, L"%a is not a full token\n", marker);
+			}
+
+			CopyMem(part1, builtin, part1_len);
+			*(part1 + part1_len) = '\0';
+			CopyMem(part2, p + marker_len, part2_len);
+			*(part2 + (part2_len)) = '\0';
 		}
 	}
 
+	// namespace appeared in the builtin
+	if (strstra(part1, namespace) != NULL || strstra(part2, namespace) != NULL) {
+		status = EFI_INVALID_PARAMETER;
+		UnicodeSPrint(errbuf, errbuf_buflen, L"%a appears in builtin cmdline\n", namespace);
+		goto out;
+	}
+
+	// namespace appears in runtime
+	if (strstra(runtime, namespace) != NULL) {
+		status = EFI_INVALID_PARAMETER;
+		UnicodeSPrint(errbuf, errbuf_buflen, L"%a appears in runtime cmdline\n", namespace);
+		goto out;
+	}
+
+	// Print(L"part1=%a|\npart2=%a|\nruntime=%a|\n", part1, part2, runtime);
+	status = check_cmdline(runtime, runtime_len, errbuf, errbuf_buflen);
+
+	// EFI_SECURITY_VIOLATION is allowed if insecure boot, so continue on.
+	if ((status == EFI_SECURITY_VIOLATION && secure) || EFI_ERROR(status)) {
+		goto out;
+	}
+
+	// At this point, part1 and part2 are set so we can just concatenate part1, runtime, part2
+	UINTN clen = part1_len + runtime_len + part2_len;
+	CHAR8 *cbuf;
+	cbuf = AllocatePool(clen + 1);
+	CopyMem(cbuf, part1, part1_len);
+	CopyMem(cbuf+part1_len, runtime, runtime_len);
+	CopyMem(cbuf+part1_len+runtime_len, part2, part2_len);
+	cbuf[clen] = '\0';
+	*cmdline = cbuf;
+	*cmdline_len = clen;
+
 out:
+	if (errbuf != NULL && errbuf[0] == '\0') {
+		FreePool(errbuf);
+		errbuf = NULL;
+	}
+
+	if (part1 != NULL)
+		FreePool(part1);
 
-	FreePool(buf);
+	if (part2 != NULL)
+		FreePool(part2);
+
+	*errmsg = errbuf;
 	return status;
 }
+
+// get_cmdline_with_print:
+//   check the command line
+//   return EFI_SUCCESS if command line can be booted.
+//
+// Note: get_cmdline (called by get_cmdline_with_print) will
+// return EFI_SECURITY_VIOLATION the same for secure and insecure.
+// if insecure, this function changes a EFI_SECURITY_VIOLATION return
+// value to EFI_SUCCESS.
+EFI_STATUS get_cmdline_with_print(
+		BOOLEAN secure,
+		CONST CHAR8 *builtin, UINTN builtin_len,
+		CONST CHAR8 *runtime, UINTN runtime_len,
+		CHAR8 **cmdline, UINTN *cmdline_len) {
+
+	CHAR16 *errmsg = NULL;
+	EFI_STATUS err;
+
+	err = get_cmdline(secure,
+		builtin, builtin_len, runtime, runtime_len,
+		cmdline, cmdline_len, &errmsg);
+
+	if (!EFI_ERROR(err)) {
+		goto out;
+		return err;
+	}
+
+	if (errmsg == NULL) {
+		Print(L"%r\n");
```

Comment:
Sorry - what is a %r?  I gather from google that it's special for EFI_STATUS, but does it need an argument or assume an internal variable as its argument?